### PR TITLE
ENH: Allow multi values for index and columns in df.pivot

### LINKED ISSUE #21425: BUG: DataFrame.pivot fails on multiple columns to set as index
I am not sure anymore if `DataFrame.pivot` actually supports multiple columns to set as `index`/`columns` (it should), but in any case this error is very confusing:

```
In [1]: df = pd.DataFrame({'lev1': [1, 1, 1, 1,2, 2, 2,2], 'lev2': [1, 1, 2, 2, 1, 1, 2, 2], 'lev3': [1, 2, 1, 2, 1, 2, 1, 2], 'values': range(8)})

In [2]: df
Out[2]: 
   lev1  lev2  lev3  values
0     1     1     1       0
1     1     1     2       1
2     1     2     1       2
3     1     2     2       3
4     2     1     1       4
5     2     1     2       5
6     2     2     1       6
7     2     2     2       7

In [3]: df.pivot(index=['lev1', 'lev2'], columns='lev3', values='values')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-2fef29f9fd39> in <module>()
----> 1 df.pivot(index=['lev1', 'lev2'], columns='lev3', values='values')

~/scipy/pandas/pandas/core/frame.py in pivot(self, index, columns, values)
   5191         """
   5192         from pandas.core.reshape.reshape import pivot
-> 5193         return pivot(self, index=index, columns=columns, values=values)
   5194 
   5195     _shared_docs['pivot_table'] = """

~/scipy/pandas/pandas/core/reshape/reshape.py in pivot(self, index, columns, values)
    406         else:
    407             indexed = self._constructor_sliced(self[values].values,
--> 408                                                index=index)
    409     return indexed.unstack(columns)
    410 

~/scipy/pandas/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    260                             'Length of passed values is {val}, '
    261                             'index implies {ind}'
--> 262                             .format(val=len(data), ind=len(index)))
    263                 except TypeError:
    264                     pass

ValueError: Length of passed values is 8, index implies 2

```

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #21425
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by charlesdong1991** [comment]
would be nice if someone could take a look here, i just rebased ^^

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -228,6 +228,7 @@ Other enhancements
 - Added an experimental :attr:`~DataFrame.attrs` for storing global metadata about a dataset (:issue:`29062`)
 - :meth:`Timestamp.fromisocalendar` is now compatible with python 3.8 and above (:issue:`28115`)
 - :meth:`DataFrame.to_pickle` and :func:`read_pickle` now accept URL (:issue:`30163`)
+- :meth:`DataFrame.pivot` can now take lists for ``index`` and ``columns`` arguments (:issue:`21425`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: move to 1.1

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5847,11 +5847,19 @@ def groupby(
 
         Parameters
         ----------%s
-        index : str or object, optional
+        index : str or object or a list of the previous, optional
```
Comment (jreback) [inline pandas/core/frame.py]: list of str, this really takes object? what exactly does that mean

**Review by jreback — CHANGES_REQUESTED** [review]
can you add an example in the docs as well

File: pandas/core/frame.py (hunk 2 above)
Comment (charlesdong1991) [reply pandas/core/frame.py]: oh, my bad, i copied the description from `values`, will update a bit to make it clearer

**Comment by charlesdong1991** [comment]
just in case if you'd like to take a look too @WillAyd thanks! ^^

File: pandas/core/frame.py [hunk 3]
```diff
@@ -5847,11 +5847,19 @@ def groupby(
 
         Parameters
         ----------%s
-        index : str or object, optional
+        index : str or object or a list of str, optional
             Column to use to make new frame's index. If None, uses
             existing index.
-        columns : str or object
+
+            .. versionchanged:: 1.0.0
```
Comment (WillAyd) [inline pandas/core/frame.py]:
```suggestion
            .. versionchanged:: 1.1.0
```

File: pandas/core/frame.py [hunk 4]
```diff
@@ -5847,11 +5847,19 @@ def groupby(
...
+               Also accept list of index names.
+
+        columns : str or object or a list of str
             Column to use to make new frame's columns.
+
+            .. versionchanged:: 1.0.0
```
Comment (WillAyd) [inline pandas/core/frame.py]:
```suggestion
            .. versionchanged:: 1.1.0
```

File: pandas/core/reshape/pivot.py [hunk 5]
```diff
@@ -431,17 +431,31 @@ def _convert_by(by):
 def pivot(data: "DataFrame", index=None, columns=None, values=None) -> "DataFrame":
     if columns is None:
         raise TypeError("pivot() missing 1 required argument: 'columns'")
+    columns = columns if is_list_like(columns) else [columns]
 
     if values is None:
-        cols = [columns] if index is None else [index, columns]
+        cols: List[str] = []
```
Comment (WillAyd) [inline pandas/core/reshape/pivot.py]: I think you want `List[Label]` (import from pandas._typing) to allow anything column name, unless this really does have to be string-only

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/reshape/pivot.py (hunk 5 above)
Comment (jreback) [reply pandas/core/reshape/pivot.py]: i don't think this will matter until we type the signature, so ok with handling this as a followon.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 3 above)
Comment (charlesdong1991) [reply pandas/core/frame.py]: nice catch!! 😅

File: pandas/core/frame.py (hunk 4 above)
Comment (charlesdong1991) [reply pandas/core/frame.py]: done!

File: pandas/tests/reshape/test_pivot.py [hunk 6]
```diff
@@ -1007,6 +1007,192 @@ def test_pivot_no_level_overlap(self):
         expected = grouped.unstack("b").unstack("c").dropna(axis=1, how="all")
         tm.assert_frame_equal(table, expected)
 
+    @pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/tests/reshape/test_pivot.py]: thanks! I think a bit, and feel maybe it's better to make this change here to make the scope of PR clearer, so i made a new file `test_pivot_multilevel.py` and moved related tests there.

**Comment by charlesdong1991** [comment]
HI, @jreback  and @WillAyd thanks for both of yours reviews, appreciate a lot!!

One thing to mention: i'd like to have signature added as a follow-up to make this clearer, since we have PRs addressing TYPING issues only if you agree. Otherwise, I think I have addressed both of your reviews in the change ^^

Feel free for further reviews! Thanks!

**Comment by charlesdong1991** [comment]
just updated with latest master

**Comment by charlesdong1991** [comment]
gentle ping @WillAyd for feedback? ^^

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

**Comment by charlesdong1991** [comment]
gentle ping @jreback ^^

**Comment by charlesdong1991** [comment]
sorry for pushing for a review, I know you are so busy with so many PRs. This PR has been rebased several times, so some feedbacks would be appreciated @jreback 👍

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @charlesdong1991 
very nice

**Pull request merged**